hnn_core/network.py

```diff
@@ -205,10 +219,22 @@ def __init__(self, params):
         self.n_cells = sum(len(self.pos_dict[src]) for src in
                            self.cellname_list)
 
-        # The legacy code in HNN-GUI _always_ defines 2 'common' and 5
-        # 'unique' feeds. They are added here for backwards compatibility
-        # until a new handling of external NetworkDrives's is completed.
-        self._add_external_feeds()
+        if initialise_hnn_drives:
+
+            self.attach_bias('tonic')  # add them if present in params
+
+            # XXX 'common' is actually two very different drives (poor naming)
+            self.attach_drive(drive_name='common', drive_params=p_common,
+                              cell_specific=False)
+
+            # count on create_pext figuring out which drives are present
+            # NB needs to be sorted alphabetically for tests to pass
+            for drive_name in sorted(p_unique.keys()):
+                self.attach_drive(drive_name=drive_name,
+                                  drive_params=p_unique[drive_name],
+                                  cell_specific=True)
+
+            self.instantiate_drives(n_trials=self.params['N_trials'])
```

Comment:
Something I was going to propose in #212 is that drive instantiation (i.e., calling the `net.instantiate_drive()` function that creates the actual event times associated with an external drive) be called once and only once: in `simulate_dipole()`. Currently, there is the possibility of running `instantiate_drives()` twice, first when instantiating a `Network` object and second when the user calls `simulate_dipole()`. This indicates to me that calling `instantiate_drives()` in the `Network` constructor is too early. To avoid spaghetti code, we should probably all come come to an agreement on where and when the specific drive event times should be created. I think it will help clarify [this](https://github.com/jonescompneurolab/hnn-core/pull/221#issuecomment-740892976) discussion.

Reply:
I agree that `instantiate_drive()` does not need to be called by `net.__init__()`! However, I envision a workflow where

1. attach drives
2. play around with timing stats/parameters, interactively setting delays and dispersions using a notebook (histograms of event times)
3. build & simulate

In the above, 2. would need to instantiate event times repeatedly, possibly with different seeds. Once 3. is reached, it should be possible to _not_ re-instantiate the event times, if the user wants to get precisely the distribution they've settled on in 2. 

I'm beginning to think there should be a `NetworkDrive` class, so that each drive could be instantiated _independently_ of the others... Spaghetti be damned ;)

Reply:
I think the `self.feed_times` can become your `NetworkDrive` class :) But currently it's just a `dict`, not sure if we need anything more sophisticated than that. We probably don't want any methods

Reply:
I'm mostly inspired by objects like the `Forward` in MNE which has a `data` key and other keys containing metadata

Reply:
> I think the `self.feed_times` can become your `NetworkDrive` class :)

I'll think about this! And look at `Forward` for inspiration too.